### ISSUE TITLE
Graphback context do not have typings for services

### ISSUE THREAD

**Issue by wtrocki** [aerogear/graphback#1637]

**Comment by wtrocki** [comment]
See https://github.com/aerogear/graphback/pull/1643
Challenge here is that we still need to generate types from somewhere. 
More deeper integration with GraphQL-CLI codegen is probably needed to enable better local development regime :D

**Comment by craicoverflow** [comment]
The only thing I can think of is by letting the SchemaCRUDPlugin generate the context file:

```ts
interface GeneratedContext {
  graphback: {
    services: {
      Note: GraphbackCRUDService
    }
  }
}
```

**Comment by wtrocki** [comment]
No need for generation. We will not be able to get service names but.. connecting CRUDservices is trivial:

`[service:string: GraphbackCRUDService]`

If we start using generics and drop graphback namespace this will be:

`resolveUser(_, _, { services:{ User: GraphbackCRUDService<User>}})`

AKA type it yourself - user will come from codegen then.

**Comment by craicoverflow** [comment]
The decomposition of the context wil be a little bit bulkier than what you proposed, as the typing needs to be decomposed separate to the context:

```ts
resolverFn: (_: any, __: any, { graphback: { services: Note } }: { graphback: { services: GraphbackCRUDService<Note> } })
```

But this is still the best solution.

**Comment by craicoverflow** [comment]
The problem with this approach is that you no longer have the complete context object as it is decomposed as a parameter.

```ts
const customResolvers = {
  Query: {
    getDraftNotes: async (_: any, __: any, { graphback: Note }: { graphback: GraphbackCRUDService<Note> }) => {
      const results = await Note.findBy({
        filter: {
          title: {
            startsWith: '[DRAFT]'
          }
        } as QueryFilter<NoteFilter>
      }, context, info); // there is no context object anymore!!

      return results.items;
    }
  }
}
```

While this is an optional parameter, and in fact currently the only method that uses it in the CRUDService is `batchLoadData`, it should still be kept.

So the user has three choices:

1. Sacrifice passing context to CRUDService for typing in the resolver function.
2. Sacrifice typings for a full context.
3. Add boilerplate to the function by doing some decomposition and casting in the function body.

Neither way is perfect and has its downfalls.

**Comment by craicoverflow** [comment]
**FOURTH APPROACH**

Another approach is that the user can define their own custom context interface to use in custom resolvers:

```ts
interface AppContext {
  graphback: {
    Note: GraphbackCRUDService<NoteFilter>
  }
}

const customResolvers = {
  Query: {
    getDraftNotes: async (_: any, __: any, context: AppContext, info) => {
      const results = await context.graphback.Note.findBy({}, context, info);

      return results.items;
    }
  }
}
```

This method would mean keeping the custom context interface synced with the schema models, but would only need to specify the models required for custom resolvers..and would mean less repetition of destructuring the context for every resolver, making it - easier to read and maintain.

**Comment by craicoverflow** [comment]
cc @wtrocki and @machi1990 for thoughts on this.

**Comment by machi1990** [comment]
> **FOURTH APPROACH**
> 
> Another approach is that the user can define their own custom context interface to use in custom resolvers:
> 
> ```ts
> interface AppContext {
>   graphback: {
>     Note: GraphbackCRUDService<NoteFilter>
>   }
> }
> 
> const customResolvers = {
>   Query: {
>     getDraftNotes: async (_: any, __: any, context: AppContext, info) => {
>       const results = await context.graphback.Note.findBy({}, context, info);
> 
>       return results.items;
>     }
>   }
> }
> ```
> 
> This method would mean keeping the custom context interface synced with the schema models, but would only need to specify the models required for custom resolvers..and would mean less repetition of destructuring the context for every resolver, making it - easier to read and maintain.

This will be done in user's code right?

**Comment by craicoverflow** [comment]
Correct, and would only need to be done if the user wants strongly typed models for custom resolvers, and would not override the GraphbackContext.

**Comment by machi1990** [comment]
> Correct, and would only need to be done if the user wants strongly typed models for custom resolvers, and would not override the GraphbackContext.

Understood. This makes a lot of sense and I'd see it as docs only kind of change.

**Comment by craicoverflow** [comment]
Yep but there is some crossover because I am deciding what the best approach is to add a custom resolver in https://github.com/aerogear/graphback/issues/1627 - so it is a recommended approach but there will also be a code example.

**Comment by machi1990** [comment]
> Yep but there is some crossover because I am deciding what the best approach is to add a custom resolver in #1627 - so it is a recommended approach but there will also be a code example.

You are right. Sure, having the that on templates will be a huge plus too.

**Comment by craicoverflow** [comment]
Resolved by #1956
